### DISCUSSION TITLE
SetPixel and GetPixel extensions to WritableBitmap?

### DISCUSSION THREAD

**Discussion by Shadowblitz16** [AvaloniaUI/Avalonia#16330] [Q&A]
Are there SetPixel and GetPixel extensions to WritableBitmap?
Specifically ones that use a 1d index and a color struct?
Also preferably with pixelFormat support

**Comment by maxkatz6** [comment]
No, and not planned for two reasons:
1. It's **slow** requiring Lock on each method call. Users will misuse this API in for-loops just like they do with System.Drawing.Bitmap SetPixel.
2. Implementation is not so trivial with more complex pixel formats.

**Reply by Shadowblitz16** [reply]
I mean what about ILockedFramebuffer?

**Reply by maxkatz6** [reply]
That would solve first problem, yes. 
It still needs a complete implementation for all supported pixel formats with unit tests.

In general, I don't think it's impossible to accept such change.

**Reply by Shadowblitz16** [reply]
Also how exactly do I assign a value to a writable bitmap?
```
public class Palette(uint colorCount, uint swapCount) : WriteableBitmap(new((int)colorCount, (int)swapCount+1), Vector.One, PixelFormat.Rgba8888, Avalonia.Platform.AlphaFormat.Premul)\
{
    private ILockedFramebuffer _locked = Lock();
    
    public void SetPixel(uint index, Color color)
    {
        unsafe
        {
            &((_locked.Address + index+0) % (Size.Width * Size.Height)) = color.R;
            &((_locked.Address + index+1) % (Size.Width * Size.Height)) = color.G;
            &((_locked.Address + index+2) % (Size.Width * Size.Height)) = color.B;
            &((_locked.Address + index+3) % (Size.Width * Size.Height)) = color.A;
        }
    }
    public Color GetPixel(uint index)
    {
        unsafe
        {
            var r = &((_locked.Address + index+0) % (Size.Width * Size.Height));
            var g = &((_locked.Address + index+1) % (Size.Width * Size.Height));
            var b = &((_locked.Address + index+2) % (Size.Width * Size.Height));
            var a = &((_locked.Address + index+3) % (Size.Width * Size.Height));
            return new Color(a, r, g, b);
        }
    }
}
```

**Reply by maxkatz6** [reply]
What do you mean "assign a value"?

>     private ILockedFramebuffer _locked = Lock();

You should NOT keep bitmap locked all the time. It will cause all sorts of problems and it likely won't be usable in the rendering. You need to lock it only when you need to write or read anything.

**Reply by Gillibald** [reply]
https://github.com/Gillibald/Avalonia/blob/fdffe1b1242798cbac96cfec41794b160c1805fc/src/Avalonia.Visuals/Platform/LockedFramebufferExtensions.cs

**Reply by Shadowblitz16** [reply]
> https://github.com/Gillibald/Avalonia/blob/fdffe1b1242798cbac96cfec41794b160c1805fc/src/Avalonia.Visuals/Platform/LockedFramebufferExtensions.cs

This uses two different formats for getting and setting pixels.

**Reply by Shadowblitz16** [reply]
> What do you mean "assign a value"?
> 
> > ```
> > private ILockedFramebuffer _locked = Lock();
> > ```
> 
> You should NOT keep bitmap locked all the time. It will cause all sorts of problems and it likely won't be usable in the rendering. You need to lock it only when you need to write or read anything.

Ok idk then.

**Comment by TobisLee** [comment]
Hey @Shadowblitz16 . Sorry to bother you after such a long time. I'd like to know how you solved this problem? Did you use the `SetPixel` in the link?

**Reply by Shadowblitz16** [reply]
> Hey @Shadowblitz16 . Sorry to bother you after such a long time. I'd like to know how you solved this problem? Did you use the `SetPixel` in the link?

I kinda just gave up

**Reply by TobisLee** [reply]
LOL. Thanks bro.
